### ISSUE TITLE
Add in-app locale switcher

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#408]
That override system settings, letting us ship more locales than the device offers.

This might *just work* with the Focus code that we stole. :)

**Comment by Delphine** [comment]
As long as we don't know the supported list of languages from Amazon, we're blocked in knowing which l10n repos to set up.
Regardless, I really think that when we have the possibility to add a language switcher in Firefox (so what we currently do with all our other Android products), we should strive to have it from the start. 

In this case, we wouldn't be tied to Amazon in terms of languages, and could target a much larger audience. To give an idea, we have 74 languages on Focus Android, amongst which 55 are 100% complete. 
We’d also know exactly which languages we’re actually distributing that way (since right now, that list seems to depend on the country you are in).
Language support is also a major differentiator with competitors: we have a lot of communities covering languages that other browsers don’t.

How hard do you think this would be to try and add it? (CC @Sdaswani so he stays in the loop)

**Comment by Delphine** [comment]
Another thing: Amazon's supported languages will likely change with time, and it's going to be hard to scale following that. Having the language switcher means we don't need to know what they are currently supporting (and would give us much more liberty, as I mentioned above)

**Comment by Sdaswani** [comment]
@Delphine I'll have the locale list by EoD tomorrow - sorry about that! As for the language switcher, it's up to @bbinto to prioritize it on our roadmap (it should be pretty easy to implement since we can embrace and extend it from Focus).

**Comment by Fjoerfoks** [comment]
Creating a language switcher will be awesome! Please don't take the underlying OS or vendor as ground rule, let the user decide which language he/she prefers. This is a great opportunity for minority languages.
The only downside might be that when shipping more locales, the size of the app might increase.
(If only we could have this on Firefox (Focus) for iOS too).

**Comment by Sdaswani** [comment]
@Delphine @boek here is the list of supported locales:

* English (U.S.)
* English (U.K.)
* French
* German
* Italian
* Spanish
* Japanese
* Portuguese (Brazil)
* English (India)
* English (Australia)
* Chinese (Simplified)

**Comment by Delphine** [comment]
@Sdaswani, @bbinto : Hmmm, are we sure this list is correct? That's in fact what I'm presented with when I log in from the US on the TV, but No-Jun seemed to suggest this varies from country to country.
Also if that's the list of supported languages from Amazon - why did they ask us for languages such as Hindi, Korean, Arabic and Russian for previous versions?
It seems something is not quite right, so would rather be extra sure about this before we go any further. Thanks!

**Comment by Sdaswani** [comment]
@Delphine this info came directly from Amazon, but I'm happy to ask a follow up - what question should I ask exactly?

**Comment by Delphine** [comment]
Ok so the confusion here may be that the person didn't understand what we were asking - and therefore only listed the languages they see listed under the Preferences of the TV itself.

So, since Amazon did initially ask us for other languages for Firefox - and since some of those aren't the ones that the TV preferences proposes - I think there are a few questions to ask:
- Does the language list that a user can choose from the TV Preferences vary at all from region to region, or country to country? (that would explain why they asked us for different languages for the launch)
- If so, what are the permissible languages third-party vendors can distribute their apps in? (so, us, in this case)
- And if not (ie the language list from preferences do *not* ever change according to a users location), it would be good to understand their rational behind requesting languages for Firefox that are not listed in the preferences of the TV (so that users can't choose from, in this case: Hindi, Korean, Arabic and Russian)
Does this make sense? 
Thanks Susheel!

**Comment by Sdaswani** [comment]
Yes, thanks @Delphine - this is perfect!

**Comment by Sdaswani** [comment]
Ok so it looks like Amazon thinks we should build an in-app switcher too! 😂

‘My guess is that we assumed the browser gave users the ability to choose their own language, instead of reading the language selected on the device. Given that the app uses the latter method, I would go with the languages [we provided].’

So @Delphine and @boek this should unblock us - we can create the l10n DB with all languages, and we can add the in app switcher to support them down the line.

**Comment by Delphine** [comment]
Thanks Susheel! Very helpful.
Until the locale switcher is in place, I would suggest to only enable the locales that a user can actually chose from - so the ones listed out by Amazon TV preferences. Otherwise, users won't be able to chose those languages, and we wouldn't be able to give additional value to the product in enlarging our user base quite yet (there's also a question of time + cost).
I'd therefore suggest that we start enabling a subset of these languages. I'll dive in the specifics in [Bug 1436856](https://bugzilla.mozilla.org/show_bug.cgi?id=1436856) and we can continue the conversation there so as not to further hijack this this issue that is initially about enabling the language switcher (sorry about that!) ^^ Thanks!

**Comment by Delphine** [comment]
@Sdaswani: I was wondering if this was on the roadmap for the next release? Thanks!

**Comment by Sdaswani** [comment]
I don’t think so - @bbinto any thoughts on prioritizing this?

**Comment by bbinto** [comment]
@Sdaswani - In order for me to prioritize, can we get an high level estimate on this, do we need UX? I assume?

**Comment by Sdaswani** [comment]
I think the UX would be fairly straightforward but it’s at least a medium, and probably a large, from the Eng + QA side.

**Comment by bbinto** [comment]
Ok, thanks, I wonder if this can be broken down in to several 2-week sprint & dot releases? I suggest scheduling this for after 3.0.

@Delphine -- I know L10n is important and I'm so happy you bring up these things. I might have to say "not now", because our main goal for H1 and FFTV is to include more features that support non-YouTube usage, e.g. Pocket Video feed. 

We can chat further if you believe adding a local switcher earlier is more important than a Pocket video feed in order for us to retain and grow our user base (sorry for the PM hat chat :)).

**Comment by Delphine** [comment]
Thanks everyone for chiming in! (and no worries for the PM hat chat Barbara - that's totally understandable! 😂)

To be very honest, I actually understood from @Sdaswani 's comment in this thread that Amazon was asking us to implement an in app-locale switcher ("Ok so it looks like Amazon thinks we should build an in-app switcher too! 😂"), which is one of the reasons I chimed in here. Also, I think I remember (probably @mcomella but not too sure anymore) saying in an email thread that this may be very easy to implement.

To give you an idea, right now we have 75 languages shipping for Focus Android thanks to the in-app switcher - which is really an incredible coverage compared to other similar apps, and a differentiator with our competitors. I definitely want to give more relevant/useful data but I'm on my way to head off to PTO now, and then we have a week long l10n workweek in Europe, so that may have to wait a bit.

I don't recall seeing a timeline for Firefox TV (I may be wrong though - looked in the GitHub repo and couldn't see a Milestones section either) - but I assume you're saying v3 will happen more or less around the end of H1, correct?

**Comment by Sdaswani** [comment]
@Delphine v3 is scheduled for early to mid-May, it depends on the UX specs we get in a few weeks. 

To be clear on what Amazon said: "My guess is that we assumed the browser gave users the ability to choose their own language." So I don't think they _need_ us to, but it's a definitely nice thing to have.

**Comment by Delphine** [comment]
Thanks for the clarification Susheel.
So I looked again at the list of countries where the Firefox TV is currently available in, and seems that the only corresponding locale that we're missing is Japanese at the moment. 
However it says that we're targeting India soon, so that's something to keep on the lookout for, since it's a HUGE market. We've got a lot of potential Indian locales we can translate into.
So I guess that for the moment this is not urgent, but when India comes up, it will be interesting to talk about adding the in-app locale switcher again.

**Comment by Sdaswani** [comment]
Pressed the wrong button sorry! 🤣

**Comment by Delphine** [comment]
Uhm, nevermind, we have Japanese... Gonna update the comment!
